pilot/pkg/model/virtualservice.go

```diff
@@ -0,0 +1,339 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"strings"
+
+	networking "istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/util/sets"
+)
+
+func mergeVirtualServicesIfNeeded(vServices []Config) (out []Config) {
+	out = make([]Config, 0, len(vServices))
+	delegatesMap := map[string]Config{}
+	// root virtualservices with delegate
+	var rootVses []Config
+
+	// 1. classify virtualservices
+	for _, vs := range vServices {
+		rule := vs.Spec.(*networking.VirtualService)
+		// it is delegate, add it to the indexer cache
+		if len(rule.Hosts) == 0 {
+			delegatesMap[key(vs.Name, vs.Namespace)] = vs
+			continue
+		}
```

Comment:
btw, you cannot have the gateways field as well in the delegate

Reply:
?mismatched line